frontend/src/router.ts

```diff
@@ -170,16 +180,46 @@ export const router = new Router({
   ]
 });
 
-router.beforeEach((to, from, next) => {
-  let currentUser = firebase.auth().currentUser;
-  let requireAuth = to.matched.some(record => record.meta.requireAuth);
-  let guestOnly = to.matched.some(record => record.meta.guestOnly);
+export const routerBeforeEach: NavigationGuard = async (to, from, next) => {
+  // reverse routes so nested routes can take control
+  const requireAuthRoute = [...to.matched]
+    .reverse()
+    .find(({ meta }) => meta.auth);
+
+  if (!requireAuthRoute) return next();
+
+  const currentUser = firebase.auth().currentUser;
+  const {
+    meta: {
+      auth: { requireAuth, guestOnly, requireClaims }
+    }
+  } = requireAuthRoute;
 
   if (requireAuth && !currentUser)
     next({
       path: "/login",
       query: { redirect: to.fullPath }
     });
   else if (guestOnly && currentUser) next("/");
-  else next();
-});
+  else if (requireClaims) {
+    const userClaims = currentUser
+      ? (await currentUser.getIdTokenResult()).claims
+      : null;
+
+    if (!userClaims) next("/restricted");
+    else {
+      const hasRequiredClaim = Object.entries(requireClaims).some(
+        ([claimName, claimValue]) => {
+          return userClaims[claimName] === claimValue;
```

Comment:
I guess we can simplify this by `([claimName, claimValue]) => userClaims[claimName] === claimValue`